torchqdynamics/odeint.py

```diff
@@ -1,80 +1,124 @@
 import warnings
+from abc import ABC, abstractmethod
+from enum import Enum
+from typing import List
 
 import torch
 
 from .solver import AdaptativeStep, FixedStep
 
 
-def odeint(qsolver, y0, tsave, sensitivity='autograd', parameters=None):
-    # check arguments
-    tsave = torch.tensor(tsave)
-    check_tsave(tsave)
-    if (parameters is not None) and (sensitivity in [None, 'autograd']):
-        warnings.warn('Argument `parameters` was supplied in `odeint` but not used.')
+class ForwardQSolver(ABC):
+    @abstractmethod
+    def forward(self, t, dt, rho):
+        pass
 
-    # dispatch to appropriate odeint subroutine
-    if sensitivity is None:
-        return odeint_inplace(qsolver, y0, tsave)
-    elif sensitivity == 'autograd':
-        return odeint_main(qsolver, y0, tsave)
-    elif sensitivity == 'adjoint':
-        return odeint_adjoint(qsolver, y0, tsave, parameters)
 
+class GradientAlgorithm(Enum):
+    AUTOGRAD = 'autograd'  # gradient computed by torch
+    NONE = 'none'  # don't compute the gradients
+    ADJOINT = 'adjoint'  # compute the gradient using the adjoint method
+
+
+def odeint(
+    qsolver,
+    y0,
+    t_save: torch.Tensor,
+    exp_ops: List[torch.Tensor] = None,
+    save_states: bool = True,
+    gradient_algorithm=GradientAlgorithm.AUTOGRAD,
```

Comment:
I don't know if `sensitivity` was a better name or if they are synonyms, maybe @gautierronan knows better?

Reply:
I have seen `sensealg` and `sensitivity`. I think they are the usual terminology. IMO `gradient_algorithm` suggests something like Adam or SGD (a gradient descent alg.). An alternative could be `autodiff_algorithm` or something of this sort.